### ISSUE TITLE
False positive for SC2034

### ISSUE THREAD

**Issue by Krasauskas** [koalaman/shellcheck#2787]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2034, SC2034
- My shellcheck version (`shellcheck --version` or 'online'): online and 0.8.0
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```bash
#!/usr/bin/env bash

set -euo pipefail
IFS=$'\n\t'

function getStatefileAge() {
     declare -ig BIRTH
     declare -i current
     BIRTH=$(stat --format='%W' "$STATEFILE")
     current=$(date +%s)
     declare -ig AGE=current-BIRTH
 }
 
 getStatefileAge
 echo "$AGE"
```

#### Here's what shellcheck currently says:
```
Line 9	SC2034: BIRTH appears unused. Verify use (or export if used externally).
Line 10	SC2034: current appears unused. Verify use (or export if used externally).
```
#### Here's what I wanted or expected to see:

No errors.
Both variables are used to declare $AGE

**Comment by wileyhy** [comment]
Is it that the variables need some adjustments?

- declare -ig AGE=current-BIRTH

+ declare -ig AGE=${current}-${BIRTH}

**Comment by wileyhy** [comment]
Would something like "SC2050 – This expression is constant. Did you forget the `$` on a variable?" have possibly been more accurate of an error message?

https://www.shellcheck.net/wiki/SC2050

**Comment by Krasauskas** [comment]
> Is it that the variables need some adjustments?
> - declare -ig AGE=current-BIRTH
> + declare -ig AGE=${current}-${BIRTH}

To the best of my understading arithmetic operations do not need the $ notation; the curly braces are only really necessary for vars within string and parameter expansion. Unless I'm missing something.

> Would something like "SC2050 – This expression is constant. Did you forget the $ on a variable?" have possibly been more accurate of an error message?

I believe current behaviour of shellcheck to be buggy, because not using the $ on variable is the correct way in bash. Same way you would unset a variable - `unset varname` without a $.

**Comment by wileyhy** [comment]
Oh, arithmetic expansion. Sorry, I should have asked first before assuming.
In that case,

> $ `foo=1`
> $ `bar=2`
> $ `baz=$(( foo + bar ))`
> $ `echo "$baz"`
> 3

**Comment by Krasauskas** [comment]
Yes, I've used the `$(( ))` notation, but the `declare sum=var1+var2` also works just fine. Shellcheck is not taking lack of `$` into account and complains about an unused variable when it is, in fact, used.

**Comment by TinCanTech** [comment]
For clarity:

    $ var1=1; var2=2; declare -i sum=var1+var2; echo $sum
    $ 3

**Comment by wileyhy** [comment]
I see. Well, I was only trying to help. Thanks for showing me that.

> ...because not using the $ on variable is the correct way in bash.
>

Sorry, but it's just _a_ correct way among numerous valid arithmetic syntaxes in bash, actually. It's probably a long standing debate, to do integer attribute declaration or not to do, and my teachers have been among those who fell on the side of "not to do," arguing against issues of portability, of maintainability, of bash's inner workings, as well as some [bugginess](https://unix.stackexchange.com/questions/381761/what-do-declare-name-and-declare-g-do). More specifically, $((...)) is Posix conformant, any future maintainers would need to know of how, unless I'm mistaken about this syntax again, integer declaration adds arithmetic to the list of expansions that occur during variable assignment, and bash stores data as strings. 

```
 $ bash -c 'f() { declare a; integer a; a=1+1; echo "$a"; }; integer() { typeset -gi "$1"; }; f'
1+1
 $ bash -c 'f() { integer a; a=1+1; echo "$a"; }; integer() { typeset -gi "$1"; }; f'
2
 ```

But now I wonder, what are some of the arguments in favor of integer declaration?

**Comment by wileyhy** [comment]
```
 $ fn() { declare    i; declare -p i; declare -gi i; declare -p i; }
 $ fn
declare -- i
declare -- i
 $ fn() { declare -I i; declare -p i; declare -gi i; declare -p i; }
 $ fn
declare -i i
declare -i i
```

**Comment by TinCanTech** [comment]
Company A: _We use `POSIX` because our code needs to be portable and maintainable._

Company B: _We use `bash` because we don't ..._

A competitive world, with two (and more) competing approaches.

**Comment by Krasauskas** [comment]
> But now I wonder, what are some of the arguments in favor of integer declaration?

Making a read-only variable would be something I use.
```
declare -ri sum=var1+var2
```
Unless you mean specifying `-i` itself, to which I would argue it's good to always be as explicit as possible.
